File: category.cpp
#include "category.h"
#include <iostream>
#include <algorithm>  // Required for std::remove

category::category(const std::string &name) : name(name) {}

void category::add_product(std::shared_ptr<product> product) {
    products.push_back(product);
}

void category::remove_product(const std::shared_ptr<product>& product) {
    auto it = std::remove(products.begin(), products.end(), product);
    if (it != products.end()) {  // Ensure the product exists before erasing
        products.erase(it, products.end());
    }
}


void category::add_subcategory(std::shared_ptr<category> subcategory) {
    subcategories.push_back(subcategory);
}

File: category.py
class Category:
    def __init__(self, name):
        self._name = name
        self._subcategories = []
        self._products = []

    def add_product(self, product):
        self._products.append(product)
        print(f"Product added to the category {self._name}: {product.get_item_name()}")

    def remove_product(self, product):
        self._products.remove(product)
        print(f"Product removed from the category {self._name}: {product.get_item_name()}")

    def add_subcategory(self, category):
        self._subcategories.append(category)
        print(f"Subcategory added to {self._name}: {category._name}")
